File: frontend/src/types/project.ts
// Project-related types

import type { BaseModel } from './base';
import type { Experiment } from './experiment';

export const ProjectStatus = {
  DRAFT: 'draft',
  ACTIVE: 'active',
  PAUSED: 'paused',
  COMPLETED: 'completed',
  ARCHIVED: 'archived',
} as const;

export type ProjectStatus = typeof ProjectStatus[keyof typeof ProjectStatus];

export interface Project extends BaseModel {
  name: string;
  description: string | null;
  problem_type: string;
  status: ProjectStatus;
  spec_file_path: string;
  experiment_count: number;
  best_score: number | null;
}

export interface ProjectDetail extends Project {
  experiments: Experiment[];
  config: Record<string, any>;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  problem_type: string;
  template?: string;
  specification: {
    file_content: string;
    evolve_function: string;
    evaluate_function: string;
  };
}

export interface UpdateProjectRequest {
  name?: string;
  description?: string;
  status?: ProjectStatus;
}

export interface ProjectListResponse {
  projects: Project[];
  total: number;
  limit: number;
  offset: number;
}

File: backend/models/project.py
"""Project database model."""

import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import Column, DateTime, Enum, String, Text, UUID
from sqlalchemy.orm import relationship

from .base import Base

if TYPE_CHECKING:
    from .experiment import Experiment


class ProjectStatus(str, enum.Enum):
    """Project status enum."""

    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class Project(Base):
    """Project model."""

    __tablename__ = "projects"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), unique=True, nullable=False, index=True)
    description = Column(Text)
    problem_type = Column(String(50), nullable=False)
    status = Column(
        Enum(ProjectStatus), nullable=False, default=ProjectStatus.DRAFT, index=True
    )
    spec_file_path = Column(String(500), nullable=False)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)
    updated_at = Column(
        DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    # Relationships
    experiments = relationship(
        "Experiment", back_populates="project", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<Project(id={self.id}, name='{self.name}', status={self.status})>"
